doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+
+   .. insertprodn ltac2_match_key ltac2_match_pattern
+
+   .. prodn::
+      ltac2_match_key ::= lazy_match!
+      | match!
+      | multi_match!
+      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
+      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
+      ltac2_match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+
+   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
+   sequentially with the :token:`ltac2_match_pattern`\s, which may have
+   metavariables.  When a match is found, metavariable values are substituted
+   into :n:`@ltac2_expr`, which is then applied.
+
+   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
+   is specified.
+
+   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
+      I don't when which is required
+
+   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
+   in the :n:`@ltac2_expr`\s, the question mark is omitted.  Choose your metavariable
+   names with care to avoid name conflicts.  For example, if you use the metavariable `S`,
+   then the :token:`ltac2_expr` can't use `S` to refer to the constructor of `nat`
+   without qualifying the constructor as `Datatypes.S`.
+
+   .. todo below: is matching non-linear unification?  is it the same or different
+      from unification elsewhere in Coq?
+
+   Matching is non-linear: if a
+   metavariable occurs more than once, each occurrence must match the same
+   expression.  Matching is first-order except on variables of the form :n:`@?@ident`
+   that occur in the head position of an application. For these variables,
+   matching is second-order and returns a functional term.
+
+   .. todo the `@?ident` form is in dangling_pattern_extension_rule, not included in the doc yet
+      maybe belongs with "Applications"
+
+   `lazy_match!`
+      Causes the match to commit to the first matching branch
+      rather than trying a new match if :n:`@ltac2_expr` fails.
+      :ref:`Example<ltac2_match_vs_lazymatch_ex>`.
+
+   `match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      Failures in subsequent tactics (after the `match!`) will not cause selection
+      of a new branch.  Examples :ref:`here<ltac2_match_vs_lazymatch_ex>` and
+      :ref:`here<ltac2_match_vs_multimatch_ex>`.
+
+   `multi_match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      When a :n:`@ltac2_expr` succeeds for a branch, subsequent failures
+      (after the `multi_match!`) causing consumption of all the successes
+      of :n:`@ltac2_expr` trigger selection of a new matching branch.
+      :ref:`Example<ltac2_match_vs_multimatch_ex>`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
```

Comment:
It can also contain `@?foo`, no?

Reply:
Yes.  Can you explain what it means?  And if so, can you provide an example?  At @Zimmi48's suggestion, I omitted the relevant production from the grammar for `term10`.  (BTW there's still a good bit of syntax that we don't explain such as in the `term*` productions.  Nor do we fully explain patterns anywhere.  We'll address these in future PRs.)

Reply:
I believe I already explained this in the Ltac1 PR, let met try to find it

Reply:
Quoting https://github.com/coq/coq/pull/12103#discussion_r436297754:
> 
> As for the second paragraph, here's the same tactic but written using second-order pattern matching:
> ```coq
> Ltac uncurry f :=
>   lazymatch (eval cbv beta in f) with
>   | fun (a : ?A) (b : ?B) => @?f a b
>     => uncurry (fun ab : A * B => f (fst ab) (snd ab))
>   | ?f => f
>   end.
> ```
> 
> Perhaps a better example of the former is swapping args:
> [...]
> But you can also do with with a neat trick of `@?` patterns:
> ```coq
> Ltac swap_args f :=
>   lazymatch f with
>   | fun a b => @?f b a (* note the reversed order *)
>     => f
>   end.
> 
> Goal True.
>   let f := swap_args (fun x y => x + y) in
>   pose f. (* fun y x : nat => x + y : nat -> nat -> nat *)
> ```

I'll let you convert these to Ltac2, let me know if you have any trouble with the conversion.

Reply:
Now I recall the Ltac1 discussion.  I will leave this for a future PR.

Reply:
Maybe just add 
```suggestion
      metavariables in the form :n:`?@ident` and :n:`@?@ident`.  :g:`_` can be used to match
```
?

Reply:
And maybe add a todo referencing the comment above?